File: frontend/packages/types/src/passport.ts
export type PassportTierCode =
  | "basic"
  | "silver"
  | "gold"
  | "neo_arrivant"
  | "press_creator"
  | "business";

export type PassportStatus = "active" | "suspended" | "revoked";

export type PassportStampSource = "organization" | "qr";

export type PassportStampKind = "visit" | "memory";

export type PartnerOfferType =
  | "drink"
  | "discount"
  | "vip"
  | "gift"
  | "event_access"
  | "custom";

export type OfferRedemptionStatus = "pending" | "completed" | "cancelled" | "failed";

export interface PassportTier {
  id: string;
  code: PassportTierCode;
  name: string;
  description: string | null;
  display_order: number;
  flags: Record<string, unknown>;
}

export interface PassportStats {
  stamps_count: number;
  redemptions_count: number;
  last_stamp_at: string | null;
}

export interface PassportProgressionHint {
  next_tier_code: PassportTierCode | null;
  next_tier_label: string | null;
  hint: string | null;
  reputation_score: number;
  points_to_next: number | null;
}

export interface PassportMe {
  id: string;
  user_id: string;
  city: string;
  passport_number: string;
  qr_token: string;
  status: PassportStatus;
  tier: PassportTier;
  stats: PassportStats;
  reputation_score?: number;
  progression?: PassportProgressionHint | null;
  tier_unlocked_at?: string | null;
  onboarding_completed: boolean;
  onboarding_step: string | null;
  activated_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface PassportActivateRequest {
  city?: string | null;
}

export interface PassportStampOrganization {
  id: string;
  slug: string;
  name: string;
  city: string;
  logo_url: string | null;
}

export interface PassportStamp {
  id: string;
  kind: PassportStampKind;
  stamped_at: string;
  passport_id?: string | null;
  organization_id?: string | null;
  stamp_source?: PassportStampSource | null;
  organization?: PassportStampOrganization | null;
  title?: string | null;
  description?: string | null;
  icon?: string | null;
  slug?: string | null;
  city?: string | null;
  human_line?: string | null;
}

export interface PassportStampListResponse {
  items: PassportStamp[];
  total: number;
}

export interface PassportTierListResponse {
  items: PassportTier[];
}

export interface PartnerOfferOrganization {
  id: string;
  slug: string;
  name: string;
  city: string;
  logo_url: string | null;
}

export interface PartnerOffer {
  id: string;
  organization_id: string;
  title: string;
  description: string | null;
  offer_type: PartnerOfferType;
  tier_code_required: string | null;
  valid_from: string | null;
  valid_until: string | null;
  is_flash?: boolean;
  flash_ends_at?: string | null;
  remaining_hours?: number | null;
  remaining_minutes?: number | null;
  organization: PartnerOfferOrganization;
}

export interface PartnerOfferListResponse {
  items: PartnerOffer[];
  total: number;
}

export interface Redemption {
  id: string;
  passport_id: string;
  partner_offer_id: string;
  status: OfferRedemptionStatus;
  redeemed_at: string | null;
  created_at: string;
}

export type PassportStampClaimOrganization = {
  id: string;
  slug: string;
  name: string;
  city: string;
  logo_url: string | null;
};

export type PassportStampClaimItem = {
  id: string;
  organization_id: string;
  organization: PassportStampClaimOrganization;
  stamp_source: PassportStampSource;
  stamped_at: string;
};

export type PassportStampClaimPassportSummary = {
  stamps_count: number;
  tier_code: string;
};

export type PassportStampClaimResult = {
  status: "created" | "already_claimed";
  already_claimed: boolean;
  stamp: PassportStampClaimItem;
  passport: PassportStampClaimPassportSummary;
};

export type PassportQrTokenResponse = {
  qr_url: string;
  expires_at: string;
};

File: backend/app/schemas/passport.py
"""Passport API schemas."""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.core.passport_constants import PassportStampSource, PassportStatus, PassportTierCode
from app.schemas.passport_level import PassportProgressionHint


class PassportActivateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    city: str | None = Field(default=None, max_length=128)


class PassportTierResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    code: PassportTierCode
    name: str
    description: str | None
    display_order: int
    flags: dict[str, Any]


class PassportStatsResponse(BaseModel):
    stamps_count: int
    redemptions_count: int
    last_stamp_at: datetime | None


class PassportMeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    city: str
    passport_number: str
    qr_token: str
    status: PassportStatus
    tier: PassportTierResponse
    stats: PassportStatsResponse
    reputation_score: int = 0
    progression: PassportProgressionHint | None = None
    tier_unlocked_at: datetime | None = None
    onboarding_completed: bool
    onboarding_step: str | None
    activated_at: datetime | None
    created_at: datetime
    updated_at: datetime


class PassportStampOrganizationSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    slug: str
    name: str
    city: str
    logo_url: str | None


class PassportStampKind(StrEnum):
    VISIT = "visit"
    MEMORY = "memory"


class PassportStampResponse(BaseModel):
    """Unified timeline item — partner visit or territorial memory (TICKET-504)."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    kind: PassportStampKind = PassportStampKind.VISIT
    stamped_at: datetime
    passport_id: UUID | None = None
    organization_id: UUID | None = None
    stamp_source: PassportStampSource | None = None
    organization: PassportStampOrganizationSummary | None = None
    title: str | None = None
    description: str | None = None
    icon: str | None = None
    slug: str | None = None
    city: str | None = None
    human_line: str | None = None


class PassportStampListResponse(BaseModel):
    items: list[PassportStampResponse]
    total: int


class PassportTierListResponse(BaseModel):
    items: list[PassportTierResponse]
